### PR TITLE
Output JUnit XML metadata for CircleCi

### LINKED ISSUE #11949: Output JUnit XML test metadata for CircleCI
# @benbraou has claimed this issue. Please give them a chance to work on this.

---

Here's a screenshot of a CircleCI build for another (private) project I'm a part of:

image

It clearly highlights the parts of the build that failed.

In contrast, see one of our builds: https://circleci.com/gh/facebook/react/8402. In this one the ./scripts/circleci/upload_build.sh step failed but you need to scroll through the output a significant amount to see that. This is in part because we have a custom entry point for all our CI steps ([scripts/circleci/test_entry_point.sh](https://github.com/facebook/react/blob/v16.2.0/scripts/circleci/test_entry_point.sh)) to make it easier for us to parallelize some build steps.

I'd like us to change that test_entry_point.sh script to output JUnit XML metadata for CircleCI as described in https://circleci.com/docs/2.0/collect-test-data/. Then I believe the CircleCI build results page will show exactly which step failed.

(Bonus: If we can show individual Jest tests that would be even cooler. But I think the most important part is breaking out jest vs flow vs prettier, etc. into separate chunks.)

### PR DESCRIPTION
**Pull request by obenbrahim**
Fixes #11949

# Summary
Currently, when the build fails, there is no way to know what happened other than going through the logs in CircleCI. 

The goal of this PR is to output JUnit XML metadata for every build step. This allows CircleCI summary page to display exactly which step has failed. 

# What this PR does:
1.  Handle Jest JUnit reports generation
   **jest-junit** reporter is used to output JUnit report reports in /junit/jest-results.xml
   :warning: `test_entry_point.sh` is calling `yarn test` several times. So, a unit test failure will be duplicated in the build summary
1. Handle ESLint JUnit reports generation
 **ESLint** script can either use the default `stylish` formatter or `junit` formatter.
 :warning: in JUnit mode, ESlint results are in XML format. Not only are they written to report fils, but also they are logged to console. This may be a noise to the build logs.

:warning:  For the tasks described below, the package `junit-report-builder` has been used. Initially, I was writing the JUnit report on my own.
1. Handle flow JUnit reports generation
  **Flow** result is treated as a single test that has a single test output
1. Handle prettier JUnit  reports generation
1. Handle check_license.sh JUnit  report generation
1. Handle check_modules.sh JUnit  report generation
1. test_print_warnings.sh
  **Nothing** to be done as this script is simply used to extract all messages from `warning()
1. Handle track_stats JUnit.sh  report generation
1. Handle build.sh JUnit.sh  report generation
1. Handle upload_build.sh JUnit  report generation
1. Handle JUnit report generation for coveralls
1. Handle JUnit report generation for Danger

### CODE REVIEW

**Comment by nhunzaker** [comment]
Hey! I know we're getting to this pretty late, but do you mind upstreaming this? I took a pass but got stuck on some of the Flow stuff and thought you might have an easier time with it.

Thanks!

**Comment by benbraou** [comment]
@nhunzaker sure. This week, I have limited access to my laptop unfortunately. I will work on it next week.

**Comment by nhunzaker** [comment]
@benbraou no worries at all! I'll check back in about two weeks, but please let me know if you have any questions!

**Comment by benbraou** [comment]
@nhunzaker thanks. I finished rebasing my branch on top of master. No questions for now 😃. I am just curious to see how you are going to validate my changes. Up until now, I was forcing a build failure and then checking that the correct message is displayed in CircleCi summary section (I was also checking the locally generated junit report)

File: scripts/circleci/test_entry_point.sh [hunk 1]
```diff
@@ -10,21 +10,21 @@ if [ $((0 % CIRCLE_NODE_TOTAL)) -eq "$CIRCLE_NODE_INDEX" ]; then
...
+  COMMANDS_TO_RUN+=("yarn test --maxWorkers=2 --testResultsProcessor=${JEST_PROCESSOR}")
   COMMANDS_TO_RUN+=('./scripts/circleci/check_license.sh')
   COMMANDS_TO_RUN+=('./scripts/circleci/check_modules.sh')
   COMMANDS_TO_RUN+=('./scripts/circleci/test_print_warnings.sh')
   COMMANDS_TO_RUN+=('./scripts/circleci/track_stats.sh')
 fi
 
 if [ $((1 % CIRCLE_NODE_TOTAL)) -eq "$CIRCLE_NODE_INDEX" ]; then
-  COMMANDS_TO_RUN+=('yarn test-prod --maxWorkers=2')
+  COMMANDS_TO_RUN+=("yarn test-prod --maxWorkers=2 --testResultsProcessor=${JEST_PROCESSOR}")
```
Comment (nhunzaker) [inline scripts/circleci/test_entry_point.sh]: Will this override the reporter on line 13?

File: scripts/circleci/test_entry_point.sh [hunk 2]
```diff
@@ -10,21 +10,21 @@ if [ $((0 % CIRCLE_NODE_TOTAL)) -eq "$CIRCLE_NODE_INDEX" ]; then
...
 fi
 
 if [ $((2 % CIRCLE_NODE_TOTAL)) -eq "$CIRCLE_NODE_INDEX" ]; then
   COMMANDS_TO_RUN+=('./scripts/circleci/build.sh')
-  COMMANDS_TO_RUN+=('yarn test-build --maxWorkers=2')
-  COMMANDS_TO_RUN+=('yarn test-build-prod --maxWorkers=2')
+  COMMANDS_TO_RUN+=("yarn test-build --maxWorkers=2 --testResultsProcessor=${JEST_PROCESSOR}")
+  COMMANDS_TO_RUN+=("yarn test-build-prod --maxWorkers=2 --testResultsProcessor=${JEST_PROCESSOR}")
```
Comment (nhunzaker) [inline scripts/circleci/test_entry_point.sh]: Same question here. Do these need to generate unique reports?

File: scripts/circleci/write_junit_report.sh [hunk 3]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+
+set -e
+
+node ./scripts/tasks/junit "$1" "$2" "$3"
```
Comment (nhunzaker) [inline scripts/circleci/write_junit_report.sh]: Could you leave a comment in this script about what these arguments are?

**Comment by nhunzaker** [comment]
> I am just curious to see how you are going to validate my changes. 

Great question. I might have missed something, but these commands need confirmed:

* [ ] build.sh
* [ ] check_license.sh
* [ ] check_modules.sh
* [ ] test_coverage.sh
* [ ] upload_build.sh
* [ ] facts-tracker/index.js
* [ ] runFlow.js
* [ ] prettier
* [ ] rollup/build.js
* [ ] rollup/validate/index.js
* [ ] danger.js
* [ ] version-check.js

We could just create a failing version of this branch to verify correctness. That'll have to happen for each of these changes.

File: scripts/circleci/test_entry_point.sh (hunk 1 above)
Comment (benbraou) [reply scripts/circleci/test_entry_point.sh]: I noticed that it does not override but rather duplicate the junit error messages displayed in CircleCi. I am removing this, and keeping the reporter only in line 13.

File: scripts/circleci/test_entry_point.sh (hunk 2 above)
Comment (benbraou) [reply scripts/circleci/test_entry_point.sh]: Indeed, no need for unique reports. I am removing this, and keeping the reporting only in line 13.

File: scripts/circleci/write_junit_report.sh (hunk 3 above)
Comment (benbraou) [reply scripts/circleci/write_junit_report.sh]: sure

**Comment by benbraou** [comment]
@nhunzaker thank you for the review. I have updated the PR with the requested changes. Please let me know if further actions are needed on my side.

**Pull request closed without merging**